### ISSUE TITLE
Calculate Proportional Amounts rounding direction

### ISSUE THREAD

**Issue by brunoguerios** [balancer/b-sdk#388]
Both [v2](https://github.com/balancer/balancer-v2-monorepo/blob/36d282374b457dddea828be7884ee0d185db06ba/pkg/pool-utils/contracts/lib/BasePoolMath.sol#L22) and [v3](https://github.com/balancer/balancer-v3-monorepo/blob/26fd3c5dcf247c17109df051e64c0b51f3f8bf9a/pkg/solidity-utils/contracts/math/BasePoolMath.sol#L48) have different `calculateProportionalAmounts` for amounts `in` or `out`.

Should we consider adapting b-sdk to follow that based on how/where `calculateProportionalAmounts` is used?

e.g. following that logic, RemoveLiquidityRecovery should round down results

**Comment by johngrantuk** [comment]
@brunoguerios - you've handled this now right?

**Comment by brunoguerios** [comment]
No. I handled cow-amm proportional math. This issue is meant to replicate the approach for v2 and v3 proportional math (i.e. the other helper method)

**Comment by brunoguerios** [comment]
Reviewing the code, there are 2 code paths that rely on `calculateProportionalAmounts` within the SDK.
1. `queryRemoveLiquidity` for v2 pools -> code is already an exact match for this ✅ 
2. `addLiquidityProportional` with non-bpt as `referenceAmountIn` -> it's not possible to guarantee an exact match for this 🚫

For this last item, the problem consists in answering: Which is the `bptAmount` input to a `queryAddLiquidityProportional` that would return `amountsIn` exact match to a given a `referenceAmountIn`?
In other words, follow these steps:
> referenceAmountIn [calculateProportionalAmounts] bptAmount [queryAddLiquidityProportional] amountsIn

Since there is no equivalent for that "round-trip" on SC implementation, we can only approximate the result, and we should round down in favor of leaving some dust behind.
That's already being done, so there's no need to change anything in the current code.

Ps: to account for that, tests have a small tolerance of 1000 wei diff.

**Comment by brunoguerios** [comment]
Reopening this because a recent report highlighted and issue that tokens with 6 decimals amplify the rounding issues and get transactions to revert
looking again at the v3 monorepo, I found out that there are some adjustments I can make to improve the math (e.g. use mulDivUp)
We might still get some wei off, but better than what it is today
I’ll add the existing scenario (and some others) as integration test to make sure this is indeed fixed for v2/v3 within the SDK

**Comment by brunoguerios** [comment]
False alarm. Issue was caused by something else. Current implementation seems to be working as expected.
